README.md
# Node.js Nextcloud Talk Bot | nctb

This is a Node.js module to interact with the [Rest-API](https://nextcloud-talk.readthedocs.io/en/latest/) provided by [Nextcloud](https://nextcloud.com)s [Talk App](https://github.com/nextcloud/spreed).

## Install

```bash
npm install nextcloud-talk-bot
```

## Usage

```js
// Import Class
const NextcloudTalkBot = require('nextlcoud-talk-bot');

// Create instance w/ bot login data
const bot = new NextcloudTalkBot({
    server: '<server_url_here>',
    user: '<bot_username_here>',
    pass: '<bot_password_here>',
    autoJoin: true
});

// Simple usage; reacts to all channels the bot joined
bot.onText('Somebody in?', x => x.reply('Sure, because you are!'));

// Join a channel and react to it's contents (just returns channel if already joined)
const boppy = bot.getChannel('<channel_id_here>');

// Reaction bases on RegExp (simple)
boppy.onText(/^#echo /, msg => msg.reply(msg.text.substr(6), false));

// See examples dir for more examples

// Need to start polling
bot.startPolling();
```
See [examples directory](https://github.com/boppy/nextcloud-talk-bot/tree/master/examples) for some examples on using this module.

### Requirements

Since Bots are not currently supported by Nextcloud (@ 2020-03-01), you have to create a user for your bot. Username and password (and nextclouds hostname for sure) have to be supplied while creating the bot instance (see below).

### Configuration

You can supply login information while creating the instance (see example above). You can also supply ENV vars (or `.env` file if configured).

```bash
#!/usr/bin/env bash

export NCTB_SERVER="https://mycloud.example.com"
export NCTB_USER="bot"
export NCTB_PASS="p@ssw0rd"
node mybot.js
```

## Origin

Basically I am trying to replace big-company-based services on my mobile with a fully integrated Nextcloud.

Since [Contacts](https://apps.nextcloud.com/apps/contacts), [Calendar](https://apps.nextcloud.com/apps/calendar), [Notes](https://apps.nextcloud.com/apps/notes), [Office-Apps](https://apps.nextcloud.com/apps/onlyoffice), etc. are stable enough, I went for the replacement of
Telegram/Hangouts/WhatsApp/<YouNameIt> with [Talk](https://apps.nextcloud.com/apps/spreed). One massive limitation was that I had chat bots present in my current messaging solution that I didn't want to miss. Their main propose was to notify me in case of infrastructure problems and to help my out with my *sieve-based neck attachment*.

So I took a look at the available Chat Bots for Telegram and came back to the node module I already used to use - [yagop's Telegram Bot API][yagop_repo]. So I implemented nctb to match some of its behaviour, as being an EventEmitter (based on [EventEmitter3](https://www.npmjs.com/package/eventemitter3)) and using `.onText()` for easy interaction. I also based much of the surroundings on yagop's Repo (like this README and other docs).

## Community

The first thank goes out to yagop and contributors for the 
[Telegram Bot API Node.js Module][yagop_repo].

We thank all the developers in the Open-Source community who continuously take their time and effort in advancing this project and all underlying projects.

As soon as there is one, I'll also be absolutely happy to thank any contributor!

## License

**The MIT License (MIT)**

Copyright © 2020 Henning Bopp <henning.bopp@gmail.com>

[example]:https://github.com/boppy/nextcloud-talk-bot/tree/master/examples
[yagop_repo]:https://github.com/yagop/node-telegram-bot-api


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

* This project adheres to [Semantic Versioning](http://semver.org/).
* Enjoy your stay.

## 0.1.0 | 2020-02-29

Added:

1. First Release
1. Active and passive channel joining
1. React to message, channel, adn react globally
1. Advanced as-reply logic




src/nextcloud-talk-bot.js
const fetch = require('node-fetch');
const EventEmitter = require('eventemitter3');
const querystring = require('querystring');
const fs = require("fs");
const dotenv = require('dotenv');

class NextcloudTalkBot extends EventEmitter {
    /**
     * @typedef {Object} Channel
     * @property {boolean} joined - Boolean Joined State
     * @property {number} lastKnownMessageId - Last known Id is Int
     * @property {channelSendText} sendText - Sends a Text
     * @property {channelOnText} onText - Trigger Function if a Text is Received
     */
    /**
     * @callback channelSendText
     * @param {string} Message
     */
    /**
     * @callback channelOnText
     * @param {RegExp|string} regxt - RegExp or String to Check
     * @param {Function} callback - Callback function
     */

    /**
     *
     * @param {*} [options]
     */
    constructor(options = {}) {
        super();
        this.debug = options.debug === true;

        this.server = options.server;
        this.user = options.user;
        this.pass = options.pass;
        this.setReadMarker = options.setReadMarker === false ? 0 : 1;

        if (!this.server || !this.user || !this.pass) {
            dotenv.config();
            this.server = this.server || process.env.NCTB_SERVER;
            this.user = this.user || process.env.NCTB_USER;
            this.pass = this.pass || process.env.NCTB_PASS;
            if (!this.server || !this.user || !this.pass) {
                throw Error('Server, User, and Pass have to be set');
            }
        }

        this.channels = {};
        this.storeAllow = true;
        this.joinOnInit = options.autoJoin;

        this._loadLastIds();
        if (options.autoJoin === true) {
            this.autoJoin();
        }
    }
    autoJoin() {
        this._log('* autojoin all known channels');
        this._request('room').then(({data, headers}) => {
            data.forEach(channel => channel.readOnly || this.joinChannel(channel.token, channel.displayName))
        }).then(() => {
            this.startPolling(true);
        });
    }

    /**
     *
     * @param {string} channelId - ChannelId to get
     * @param {boolean=true} [join] - If to join
     * @returns {Channel}
     */
    getChannel(channelId, join) {
        join !== false && (this.inChannel(channelId) || this.joinChannel(channelId));
        if (!this.channels.hasOwnProperty(channelId)) {
            /** @type Channel */
            this.channels[channelId] = {
                joined: false,
                lastKnownMessageId: 0,
                sendText: (text) => this.sendText(text, channelId),
                onText: (regxt, callback) => {
                    return this.onText(regxt, callback, channelId);
                }
            }
        }
        return this.channels[channelId];
    }

    /**
     *
     * @param {string} channelId - ChannelId to Check
     * @returns {boolean}
     */
    inChannel(channelId) {
        return this.channels.hasOwnProperty(channelId) && this.channels[channelId].joined;
    }

    /**
     *
     * @param {string} channelId - ChannelId to Join
     * @param {string} [name] - Optional Name of the Channel
     * @returns {boolean}
     */
    joinChannel(channelId, name) {
        const channel = this.getChannel(channelId, false);
        if (!channel.joined) {
            channel.joined = true;
            this._log(`+ [self@${channelId}] joined ${name || ''}`);
            return true;
        }
        this._log(`± [self@${channelId}] already present`);
        return false;
    }

    /**
     *
     * @param {RegExp|string} regxt - RegExp or String to Check
     * @param {Function} callback - Callback Function to Call on Match
     * @param {string} [channelId] - Optional ChannelId to React on
     * @returns {NextcloudTalkBot}
     */
    onText(regxt, callback, channelId) {
        this._log(`@ [self@${channelId}] adding callback`);
        this.on('message', message => {
            if (!channelId || message.token === channelId) {
                const text = message.message;
                if (text === regxt) {
                    callback(message);
                } else if (typeof regxt.exec === 'function' && regxt.test(text)) {
                    const match = text.match(regxt);
                    let groups = Object.values(match.groups||{});
                    groups.push(...match);
                    message.matches = match;
                    callback(message, ...groups);
                }
            }
        });
        return this;
    }

    /**
     *
     * @param {string} text - Text to Send
     * @param {string} channelId - ChannelId to Post to
     * @param {int} [messageId] - Optional MessageId to answer to
     * @returns {fetch}
     */
    sendText(text, channelId, messageId) {
        this._log(`> [self@${channelId}/${messageId || ''}] ${text}`);
        return this._request('chat/' + channelId, {
            method: 'POST',
            body: {
                message: text,
                replyTo: messageId || 0,
            },
        }).catch($ => $);
    }
    startPolling(isAutoJoinRequest) {
        if(this.joinOnInit && isAutoJoinRequest !== true){
            this._log('° [self@sys] not starting polling. waiting for autoJoin to finish');
            return;
        }
        this._log('° [self@sys] start polling');
        Object.keys(this.channels).forEach(this._poll.bind(this));
    }
    _storeLastIds() {
        if (this.storeAllow) {
            this.__storeLastIds(false);
            this.storeAllow = false;
            setTimeout(() => this.storeAllow = true, 10000);
        } else if(!this._allowStoreTimeout) {
            this._allowStoreTimeout = setTimeout(() => {
                this._allowStoreTimeout = false;
                this._storeLastIds();
            }, 10000);
        }
    }


    _buildURL(path) {
        return `${this.server}/ocs/v2.php/apps/spreed/api/v1/${path}`;
    }
    _loadLastIds() {
        try {
            const lastIds = JSON.parse(fs.readFileSync('lastIds.json').toString());
            Object.entries(lastIds).forEach(([channelId, lastKnownMessageId]) => {
                this.getChannel(channelId, false).lastKnownMessageId = lastKnownMessageId;
            });
        } catch (e) {
        }
    }
    _log(...args) {
        if (this.debug === true) {
            console.log(...args);
        } else if (typeof args[0] === 'string') {
            process.stdout.write(args[0][0]);
        } else {
            process.stdout.write('?');
        }
    }
    _parseMessage(message) {
        if (!Array.isArray(message.messageParameters)) {
            const parm = message.messageParameters;
            return message.message.replace(/{(\S+)}/, (match, prm) => {
                if (parm[prm] && parm[prm].type === 'user') {
                    return parm[prm].id;
                }
                return match;
            });
        }
        return message.message || '';
    }
    _poll(channelId) {
        const channel = this.getChannel(channelId, false);
        if (!channel.joined) {
            this._log(`∞ [self@${channelId}] Have not joined this channel. waiting a minute...`);
            setTimeout(this._poll.bind(this, channelId), 60000);
            return;
        }
        this._log(`1 [self@${channelId}] Send poll Request`);
        this._request('chat/' + channelId, {
            query: {
                lookIntoFuture: 1,
                setReadMarker: this.setReadMarker,
                lastKnownMessageId: channel.lastKnownMessageId,
            }
        }).then((data) => {
            this._log(`2 [self@${channelId}] Receives poll Request`);
            if (data.type === 'unchanged') {
                this._log(`∞ [self@${channelId}] l∞∞∞ping!`);
            } else if (data.type === 'response') {
                const messages = data.data;
                const headers = data.headers;
                if (Array.isArray(messages)) {
                    messages.forEach(message => {
                        if (message.actorId === this.user) {
                            this._log(`= [self@${channelId}] ${message.message}`);
                        } else if (message.messageType === 'system') {
                            message.text = this._parseMessage(message);
                            this._log(`§ [${message.actorId}@${channelId}] ${message.text}`);
                            this.emit('system', message);
                        } else if (message.messageType === 'comment') {
                            message.text = this._parseMessage(message);
                            this._log(`< [${message.actorId}@${channelId}] ${message.text}`);

                            /**
                             *
                             * @param {string} text - Text to send
                             * @param {boolean|int} [asReply] - Defaults to true. Sends Text as Reply (with quote)
                             *                                  If int: Sends only if X new messages in between
                             * @returns {fetch}
                             */
                            message.reply = (text, asReply) => {
                                if (message.isReplyable && asReply !== false) {
                                    if (Number.isSafeInteger(asReply)) {
                                        const diff = channel.lastKnownMessageId - message.id;
                                        return this.sendText(text, message.token, diff >= asReply ? message.id : 0);
                                    } else {
                                        return this.sendText(text, message.token, message.id);
                                    }
                                }
                                return this.sendText(text, message.token);
                            };
                            this.emit('message', message);
                        }
                    });
                }
                if (headers && headers.get('x-chat-last-given')) {
                    channel.lastKnownMessageId = +headers.get('x-chat-last-given');
                    this._storeLastIds();
                }
            } else {
                this._log('!', data);
            }
            setTimeout(this._poll.bind(this, channelId), 50);
        }).catch(error => {
            console.error(`! [${channelId}] error`, error);
            setTimeout(this._poll.bind(this, channelId), 1000);
        });
    }
    _request(path, opt = {}) {
        let url = this._buildURL(path);
        if (opt.query) {
            url += (url.includes('?') ? '&' : '?') + querystring.stringify(opt.query);
        }
        let options = {
            method: opt.method || 'GET',
        };
        options.headers = {
            'OCS-APIRequest': 'true',
            'Accept': 'application/json',
            'Authorization': 'Basic ' + Buffer.from(this.user + ":" + this.pass).toString('base64'),
        };
        if (opt.method !== 'GET' && opt.body) {
            options.headers['Content-Type'] = 'application/json';
            options.body = JSON.stringify(opt.body);
        }

        return new Promise((ok, err) => {
            fetch(url, options).then(async (res) => {
                if (res.ok) {
                    const json = await res.json();
                    let data = json.ocs;
                    if (data.meta.status === 'ok') {
                        ok({
                            type: 'response',
                            data: data.data,
                            headers: res.headers
                        });
                    } else {
                        ok({
                            type: 'error',
                            response: json
                        })
                    }
                } else if (res.status === 304) {
                    ok({
                        type: 'unchanged'
                    });
                } else {
                    err('nope!', res);
                }
            })
        });
    }
    __storeLastIds(blocking) {
        let lastIds = {};
        Object.entries(this.channels).forEach(([channelId, channel]) => {
            lastIds[channelId] = channel.lastKnownMessageId;
        });
        const func = blocking === true ? 'writeFileSync' : 'writeFile';
        return fs[func]('lastIds.json', JSON.stringify(lastIds), "utf8", $ => $);
    }
}

module.exports = NextcloudTalkBot;

examples/30-full-bot.js
const Nextcloudbot = require('../src/nextcloud-talk-bot');

const bot = new Nextcloudbot({
    debug: true,
    autoJoin: true,
});

// Simple usage; reacts to all channels the bot joined
bot.onText('Somebody in?', x => x.reply('Sure, because you are!'));

// Join a channel and react to its contents
const boppy = bot.getChannel('dy3ennzq');

// Reaction bases on RegExp (simple)
boppy.onText(/^#echo /, msg => msg.reply(msg.text.substr(6), false));

// Reaction based on RegExp (anvanced)
boppy.onText(/^#answer ((['"])(?<q1>.*?)(\2)|(?<q2>[^"']\S*)) (?<a>.*)$/, (msg, q1, q2, a) => {
    bot.onText(q1 || q2, (message) => {
        message.reply(a, false);
        msg.reply(`Answered to @${message.actorId} in #${message.token}`);
    });
});

// Let your bot execute binaries on your system... That's totally stupid. But working... ¯\_(ツ)_/¯
const {exec: execAsyc} = require('child_process');
boppy.onText(/^#exec (?<exec>.*)$/, (msg, exec) => {
    execAsyc(exec, (err, ret) => msg.reply(ret, 1));
});

boppy.onText('uptime', (msg) => {
    execAsyc('uptime', (err, ret) => msg.reply(ret, 1));
});


// Post a reply delayed by a number of seconds (like a countdown/timer/reminder)
boppy.onText(/^#timer (?<sec>\d+|(\d+[smh])+) (?<message>.*)$/, (msg, sec, message) => {
    if (isNaN(+sec)) {
        sec = sec.match(/\d+[smh]/g).map((nunit) => {
            let n = +nunit.substr(0, nunit.length - 1);
            const unit = nunit.substr(-1);
            switch (unit) {
                case 'h':
                    n *= 60;
                case 'm':
                    n *= 60;
                default:
                    n *= 1;
            }
            return parseInt(n);
        }).reduce((a, b) => a + b, 0);
    }
    msg.reply(`Resending message "${message}" in ${sec} seconds.`, false);
    setTimeout(() => msg.reply(message, 2), sec * 1000);
});

// Lets you quit your bot by command.
let allowToDie = false;
boppy.onText('#die', msg => {
    if (allowToDie) {
        bot._log('X [self] Dieing now...');
        msg.reply("ok... See you soon(ly)").then(() => {
            bot._storeLastIds(true);
            process.exit();
        });
    } else {
        bot._log('x [self] Dieing request received');
        msg.reply("o'rly? Then type #die again.");
    }
    allowToDie = true;
});


const htmlDecode = require('js-htmlencode').htmlDecode;
const querystring = require('querystring');
const fetch = require('node-fetch');
boppy.onText(/^#wiki(?: (?<language>|fr|de|es|ja|ru|it|zh|pt|ar|fa|pl|nl|id|uk|he|sv|cs|ko|vi|ca|no|fi|hu|th|el|hi|bn|ceb|tr|ro|sw|kk|da|eo|sr|lt|sk|bg|min|sl|eu|et|hr|te|nn|gl|simple|ms|bs|ka|is|sq|la|az|mk|mr|sh|af|tl|cy|lv|ta|jv|be|ast|zh-yue|ur|ga|hy|kn|ml|ne|uz|pa|my|arz|ckb|war|vo|lmo|new|ht|bpy|lb|br|tg|io|pms|su|oc|nap|nds|scn|sa|ku|wa|bar|an|ksh|szl|fy|frr|als|ia|yi|mg|km|ce|roa-tara|am|roa-rup|map-bms|bh|bcl|co|cv|dv|nds-nl|fo|fur|gan|glk|gu|ilo|pam|csb|lij|li|gv|mi|mt|nah|nrm|se|gd|nov|qu|os|pag|ps|pdc|rm|bat-smg|sco|sc|si|tt|tk|hsb|vec|fiu-vro|wuu|vls|yo|diq|zh-min-nan|zh-classical|frp|lad|kw|mn|haw|ang|ln|ie|wo|tpi|ty|crh|nv|jbo|ay|zea|eml|ky|ig|or|cbk-zam|kg|arc|rmy|ab|gn|so|kab|ug|stq|udm|ext|mzn|pap|cu|sah|tet|sd|pcd|as|sn|lo|ba|pnb|na|got|bo|dsb|chr|cdo|hak|om|sm|ee|ti|av|zu|pnt|cr|pih|ss|bi|rw|ch|xh|kl|bug|ts|tn|kv|tum|xal|st|tw|bxr|ak|lbe|za|ff|lg|ha|sg|rn|mwl|xmf|lez|bjn|mai|gom|lrc|tyv|vep|nso|kbd|ltg|rue|pfl|gag|koi|mrj|mhr|krc|ace|hif|olo|kaa|mdf|myv|azb|ady|jam|tcy|dty|atj|kbp|din|lfn|gor|inh|sat|ban|be-x-old|pi|to|ks|iu|bm|ve|ik|dz|ny|fj|ki|chy|srn))? (?<query>.*)$/, (msg, lang, query) => {
    const domain = 'https://' + (lang ? lang : 'en') + '.wikipedia.org/';
    const parms = {
        action: 'query',
        format: 'json',
        list: 'search',
        srlimit: 3,
        srsearch: query,
        srprop: 'snippet'
    };
    const uri = domain + 'w/api.php?' + querystring.stringify(parms);
    fetch(uri)
        .then(res => res.json())
        .then(json => {
            const res = json.query.search.map(res => {
                return [
                    `[${res.title}](${domain}wiki/${escape(res.title)})`,
                    htmlDecode(res.snippet.replace(/<.*?>/g, ''))
                ].join("\n");
            });
            msg.reply(`Found ${json.query.searchinfo.totalhits} hits on wikipedia.org. First 3:\n* ` + res.join(`\n* `));
        }).catch(e => {
        msg.reply(`¯\\_(ツ)_/¯ ${e.message}`);
    })
    ;
});


// Let your bot calc the sh*t out of your input. I can also do unit conversion and stuff...
//   see https://www.npmjs.com/package/mathjs
//
// simple things like
//   #calc 42-13/37+42
//   #calc 12.7 cm to inch

// variables and functions
//   #calc nine = 20 - 1
//   #calc nine = nine - 10
//   #calc myexp(x, y) = x^y
//   #calc myexp(nine, 2)

// Math mode. Lines beginning with | will not reply, starting with # exits mathmode
//   #mathmode
//   |x = 8^2
//   x-4
//   x-nine

// Important Note:
// In this implementation ALL channels share ONE instance of the mathjs library (${parser})!
// This will also share symbols, variables and functions!

const math = require('mathjs');
const parser = math.parser();

boppy.onText(/^#calc (?<calc>.*)$/, (msg, calc) => {
    let res;
    try {
        res = parser.evaluate(calc) + "";
        if (res.substr(0, 9) === 'function ') {
            res = '[CALC] function ' + calc.substr(0, msg.text.indexOf('(') - 1);
        }
    } catch (e) {
        res = "[CALC] ¯\\_(ツ)_/¯  " + e.message;
    }
    msg.reply(res);
});


let inMathMode = false;
boppy.onText(/^(?<hash>#)?(?<pipe>\|)?(?<calc>.*)$/, (msg, hash, pipe, calc) => {
    if (hash) {
        if (inMathMode) {
            msg.reply('[MATHMODE] exiting Math Mode', false);
            inMathMode = false;
        } else if (msg.text === '#mathmode') {
            msg.reply('[MATHMODE] entering Math Mode', false);
            inMathMode = true;
        } else {
            inMathMode = false;
        }
        return;
    }
    if (!inMathMode) return;
    let res;
    try {
        res = parser.evaluate(calc) + "";
        if (res.substr(0, 9) === 'function ') {
            res = '[MATHMODE] function ' + calc.substr(0, msg.text.indexOf('(') - 1);
        }
    } catch (e) {
        res = "[MATHMODE] ¯\\_(ツ)_/¯  " + e.message;
        pipe = false;
    }
    pipe || msg.reply(res, false);
});

bot.startPolling();

/*
Exampel messages
Message {
  id: 30,
  token: 'dy3ennzq',
  actorType: 'users',
  actorId: 'bopp',
  actorDisplayName: 'Henning Bopp',
  timestamp: 1582576875,
  message: 'hallu?',
  messageParameters: [],
  systemMessage: '',
  messageType: 'comment',
  isReplyable: true
}
  id: 29,
  token: 'dy3ennzq',
  actorType: 'users',
  actorId: 'bopp',
  actorDisplayName: 'Henning Bopp',
  timestamp: 1582576869,
  message: '{actor} created the conversation',
  messageParameters: { actor: { type: 'user', id: 'bopp', name: 'Henning Bopp' } },
  systemMessage: 'conversation_created',
  messageType: 'system',
  isReplyable: false

 */